diff --git a/net/slirp.c b/net/slirp.c
index 318a26e..38f6153 100644
--- a/net/slirp.c
+++ b/net/slirp.c
@@ -152,10 +152,11 @@ static int net_slirp_init(NetClientState *peer, const char *model,
                           bool ipv6, const char *vprefix6, int vprefix6_len,
                           const char *vhost6,
                           const char *vhostname, const char *tftp_export,
-                          const char *bootfile, const char *vdhcp_start,
-                          const char *vnameserver, const char *vnameserver6,
-                          const char *smb_export, const char *vsmbserver,
-                          const char **dnssearch, Error **errp)
+                          const char *bootfile, const char *rootpath,
+                          const char *vdhcp_start, const char *vnameserver,
+                          const char *vnameserver6, const char *smb_export,
+                          const char *vsmbserver, const char **dnssearch,
+                          Error **errp)
 {
     /* default settings according to historic slirp */
     struct in_addr net  = { .s_addr = htonl(0x0a000200) }; /* 10.0.2.0 */
@@ -368,8 +369,8 @@ static int net_slirp_init(NetClientState *peer, const char *model,
 
     s->slirp = slirp_init(restricted, ipv4, net, mask, host,
                           ipv6, ip6_prefix, vprefix6_len, ip6_host,
-                          vhostname, tftp_export, bootfile, dhcp,
-                          dns, ip6_dns, dnssearch, s);
+                          vhostname, tftp_export, bootfile, rootpath,
+                          dhcp, dns, ip6_dns, dnssearch, s);
     QTAILQ_INSERT_TAIL(&slirp_stacks, s, entry);
 
     for (config = slirp_configs; config; config = config->next) {
@@ -941,9 +942,9 @@ int net_init_slirp(const Netdev *netdev, const char *name,
                          ipv4, vnet, user->host,
                          ipv6, user->ipv6_prefix, user->ipv6_prefixlen,
                          user->ipv6_host, user->hostname, user->tftp,
-                         user->bootfile, user->dhcpstart,
-                         user->dns, user->ipv6_dns, user->smb,
-                         user->smbserver, dnssearch, errp);
+                         user->bootfile, user->rootpath,
+                         user->dhcpstart, user->dns, user->ipv6_dns,
+                         user->smb, user->smbserver, dnssearch, errp);
 
     while (slirp_configs) {
         config = slirp_configs;
diff --git a/qapi/net.json b/qapi/net.json
index 4beff5d..8d022b6 100644
--- a/qapi/net.json
+++ b/qapi/net.json
@@ -152,6 +152,8 @@
 #
 # @bootfile: BOOTP filename, for use with tftp=
 #
+# @rootpath: DHCP root-path option, for use with tftp=
+#
 # @dhcpstart: the first of the 16 IPs the built-in DHCP server can
 #             assign
 #
@@ -194,6 +196,7 @@
     '*host':      'str',
     '*tftp':      'str',
     '*bootfile':  'str',
+    '*rootpath':  'str',
     '*dhcpstart': 'str',
     '*dns':       'str',
     '*dnssearch': ['String'],
diff --git a/qemu-options.hx b/qemu-options.hx
index 57f2c6a..bee2263 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -1920,7 +1920,7 @@ DEF("netdev", HAS_ARG, QEMU_OPTION_netdev,
     "         [,ipv6[=on|off]][,ipv6-net=addr[/int]][,ipv6-host=addr]\n"
     "         [,restrict=on|off][,hostname=host][,dhcpstart=addr]\n"
     "         [,dns=addr][,ipv6-dns=addr][,dnssearch=domain][,tftp=dir]\n"
-    "         [,bootfile=f][,hostfwd=rule][,guestfwd=rule]"
+    "         [,bootfile=f][,rootpath=p][,hostfwd=rule][,guestfwd=rule]"
 #ifndef _WIN32
                                              "[,smb=dir[,smbserver=addr]]\n"
 #endif
@@ -2143,6 +2143,16 @@ Example (using pxelinux):
 qemu-system-i386 -hda linux.img -boot n -net user,tftp=/path/to/tftp/files,bootfile=/pxelinux.0
 @end example
 
+@item rootpath=@var{path}
+When using the user mode network stack, broadcast @var{path} as the BOOTP
+root-path. In conjunction with @option{tftp}, this can be used to network boot
+a guest from a local directory.
+
+Example (using pxelinux):
+@example
+qemu-system-i386 -hda linux.img -boot n -net user,tftp=/path/to/tftp/files,rootpath=/nfs
+@end example
+
 @item smb=@var{dir}[,smbserver=@var{addr}]
 When using the user mode network stack, activate a built-in SMB
 server so that Windows OSes can access to the host files in @file{@var{dir}}
diff --git a/slirp/bootp.c b/slirp/bootp.c
index 5dd1a41..9054023 100644
--- a/slirp/bootp.c
+++ b/slirp/bootp.c
@@ -309,6 +309,14 @@ static void bootp_reply(Slirp *slirp, const struct bootp_t *bp)
                 q += val;
             }
         }
+
+        if (slirp->root_path) {
+            val = strlen(slirp->root_path);
+            *q++ = RFC1533_ROOTPATH;
+            *q++ = val;
+            memcpy(q, slirp->root_path, val);
+            q += val;
+        }
     } else {
         static const char nak_msg[] = "requested address not available";
 
diff --git a/slirp/libslirp.h b/slirp/libslirp.h
index f90f0f5..91b9699 100644
--- a/slirp/libslirp.h
+++ b/slirp/libslirp.h
@@ -15,9 +15,9 @@ Slirp *slirp_init(int restricted, bool in_enabled, struct in_addr vnetwork,
                   struct in6_addr vprefix_addr6, uint8_t vprefix_len,
                   struct in6_addr vhost6, const char *vhostname,
                   const char *tftp_path, const char *bootfile,
-                  struct in_addr vdhcp_start, struct in_addr vnameserver,
-                  struct in6_addr vnameserver6, const char **vdnssearch,
-                  void *opaque);
+                  const char *rootpath,  struct in_addr vdhcp_start,
+                  struct in_addr vnameserver, struct in6_addr vnameserver6,
+                  const char **vdnssearch, void *opaque);
 void slirp_cleanup(Slirp *slirp);
 
 void slirp_pollfds_fill(GArray *pollfds, uint32_t *timeout);
diff --git a/slirp/slirp.c b/slirp/slirp.c
index 1cb6b07..6c0b74d 100644
--- a/slirp/slirp.c
+++ b/slirp/slirp.c
@@ -284,9 +284,9 @@ Slirp *slirp_init(int restricted, bool in_enabled, struct in_addr vnetwork,
                   struct in6_addr vprefix_addr6, uint8_t vprefix_len,
                   struct in6_addr vhost6, const char *vhostname,
                   const char *tftp_path, const char *bootfile,
-                  struct in_addr vdhcp_start, struct in_addr vnameserver,
-                  struct in6_addr vnameserver6, const char **vdnssearch,
-                  void *opaque)
+                  const char *rootpath, struct in_addr vdhcp_start,
+                  struct in_addr vnameserver, struct in6_addr vnameserver6,
+                  const char **vdnssearch, void *opaque)
 {
     Slirp *slirp = g_malloc0(sizeof(Slirp));
 
@@ -317,6 +317,7 @@ Slirp *slirp_init(int restricted, bool in_enabled, struct in_addr vnetwork,
     }
     slirp->tftp_prefix = g_strdup(tftp_path);
     slirp->bootp_filename = g_strdup(bootfile);
+    slirp->root_path = g_strdup(rootpath);
     slirp->vdhcp_startaddr = vdhcp_start;
     slirp->vnameserver_addr = vnameserver;
     slirp->vnameserver_addr6 = vnameserver6;
diff --git a/slirp/slirp.h b/slirp/slirp.h
index 898ec95..d0c4823 100644
--- a/slirp/slirp.h
+++ b/slirp/slirp.h
@@ -192,6 +192,7 @@ struct Slirp {
     /* bootp/dhcp states */
     BOOTPClient bootp_clients[NB_BOOTP_CLIENTS];
     char *bootp_filename;
+    char *root_path;
     size_t vdnssearch_len;
     uint8_t *vdnssearch;
 
